python/lsst/daf/butler/core/datastoreCacheManager.py

```diff
@@ -653,6 +682,57 @@ def scan_cache(self) -> None:
             for path_in_cache in missing:
                 self._cache_entries.pop(path_in_cache)
 
+    def known_to_cache(self, ref: DatasetRef, extension: Optional[str] = None) -> bool:
+        """Report if the dataset is known to the cache.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Dataset to check for in the cache.
+        extension : `str`, optional
+            File extension expected. Should include the leading "``.``".
+            If `None` the extension is ignored and the dataset ID alone is
+            used to check in the cache. The extension must be defined if
+            a specific component is being checked.
+
+        Returns
+        -------
+        known : `bool`
+            Returns `True` if the dataset is currently known to the cache
+            and `False` otherwise. If the dataset refers to a component and
+            an extension is given then only that component is checked.
+
+        Notes
+        -----
+        This method can only report if the dataset is known to the cache
+        in this specific instant and does not indicate whether the file
+        can be read from the cache later. `find_in_cache()` should be called
+        if the cached file is to be used.
+
+        This method does not force the cache to be re-scanned and so can miss
+        cached datasets that have recently been written by other processes.
+        """
+        if self._cache_directory is None:
+            return False
+        if self.cache_size == 0:
+            return False
+
+        if extension is None:
+            # Look solely for matching dataset ref ID and not specific
+            # components.
+            for entry in self._cache_entries.values():
+                if ref.id == entry.ref:
+                    return True
```

Comment:
Yes. With a few thousand files in the cache and 100k files to check existence for it could get annoying.

Reply:
We can't index by datasetId because there are disassembled composites so one datasetRef can be multiple files. The uniqueness is therefore the tuple of datasetId + component (which is effectively what is encoded in the file name).

Reply:
I've added an extra dict into `CacheRegistry` to simplify this.